src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -515,46 +499,102 @@ public void ClearValidationState(string key)
             }
         }
 
-        private ModelStateEntry GetModelStateForKey(string key)
+        private ModelStateNode GetNode(string key) => GetNode(key, createIfNotExists: false);
+
+        private ModelStateNode GetOrAddNode(string key) => GetNode(key, createIfNotExists: true);
+
+        private ModelStateNode GetNode(string key, bool createIfNotExists)
         {
-            if (key == null)
+            Debug.Assert(key != null);
+            if (key.Length == 0)
             {
-                throw new ArgumentNullException(nameof(key));
+                return _root;
             }
 
-            ModelStateEntry entry;
-            if (!TryGetValue(key, out entry))
+            // For a key of the format, foo.bar[0].baz["qux"] we'll create the following nodes:
+            // foo, bar, [0], baz, ["qux"]
+
+            var current = _root;
+            var previousIndex = 0;
+            int index;
+            while ((index = key.IndexOfAny(Delimiters, previousIndex)) != -1)
             {
-                entry = new ModelStateEntry();
-                this[key] = entry;
+                var keyStart = previousIndex == 0 || key[previousIndex - 1] == '.'
```

Comment:
Made an update to how we store the keys. In the previous iteration, we kept the '.' prefix. So property accessor nodes would appear as `.foo'. This made it tricky to search properties by name. I drop the `.` prefix now as part of key creation, so  `foo.bar[0].baz["qux"].property` will appear as `foo`, `bar`, `[0]`, `baz`, `["qux"]`, `property`